go/registry: Support changing a runtime's owner

ptrus opened the pull request:
This existing code ensures that the new entity actually exists and has the required stake:
https://github.com/oasisprotocol/oasis-core/blob/2680a6e97f0ebf2ffa892c72cbaf7289e85f3938/go/consensus/tendermint/apps/registry/transactions.go#L749-L775

However, in the existing registry node tests `params.DebugBypassStake` is used, so this cannot really be tested there. That's why I added a separate unit test suite for register runtime (with is currently only testing the updating entity owner behavior).

kostko reviewed (commented):
Hm, so one thing to take care is removing the stake claim for the old entity which currently doesn't seem to happen?

Also this should be caught by the sanity checks in tests, so this needs to be tested somewhere where `DebugBypassStake` is not set.

ptrus commented:
> Hm, so one thing to take care is removing the stake claim for the old entity which currently doesn't seem to happen?

Fixed.

> Also this should be caught by the sanity checks in tests, so this needs to be tested somewhere where DebugBypassStake is not set.

Added this to the txsoruce test (workload/registration). I also tested this without the fix for removing stake claim and confirmed that sanity checker detects it.

kostko commented on go/consensus/tendermint/apps/registry/transactions.go:
```diff
@@ -770,6 +772,18 @@ func (app *registryApplication) registerRuntime( // nolint: gocyclo
 			return nil, err
 		}
+
+		// If owner was updated remove stake claim from previous owner.
```
Actually, this should also happen when changing the governance model to runtime as the `acctAddr` changed? I guess this transition is never tested in E2E tests as it should otherwise cause the sanity check to fail? Also why isn't the above just using `rt.StakingAddress()` to compute `acctAddr`?

ptrus replied:
> Actually, this should also happen when changing the governance model to runtime as the acctAddr changed

Added

> Also why isn't the above just using rt.StakingAddress() to compute acctAddr

Refactored this whole section to make it cleaner.

> I guess this transition is never tested in E2E tests

Yeah never tested. I'll look if there's an easy way to add it.

ptrus replied:
Added a test for this to `runtime-dynamic` E2E test.

kostko commented on go/consensus/tendermint/apps/registry/transactions.go:
```diff
@@ -746,30 +748,31 @@ func (app *registryApplication) registerRuntime( // nolint: gocyclo
+
+		// If owner was updated remove stake claim from previous owner.
+		if existingRt != nil && *existingRt.StakingAddress() != *rtAddress {
+			if err = stakingState.RemoveStakeClaim(ctx, *existingRt.StakingAddress(), claim); err != nil {
```
Given this is now doing multiple operations related to stake claims, it would be better to do the updates inside a transaction similar to how it is done in `registerNode`:
```golang
	// Start a new transaction and rollback in case we fail.
	ctx = ctx.NewTransaction()
	defer ctx.Close()

	// ...

	ctx.Commit()
```

ptrus replied:
Added.

kostko commented on go/oasis-node/cmd/debug/txsource/workload/registration.go:
```diff
@@ -290,13 +303,22 @@ func (r *registration) Run( // nolint: gocyclo
+			cancel()
+		}
+		loopCtx, cancel = context.WithTimeout(ctx, registryIterationTimeout)
+		defer cancel()
```
Add empty line below.

kostko commented on go/oasis-node/cmd/debug/txsource/workload/registration.go:
```diff
@@ -325,6 +347,30 @@ func (r *registration) Run( // nolint: gocyclo
 		)
 
+		// Periodically re-register the runtime with a new owner.
+		if iteration&20 == 0 {
```
Maybe move this "owner change interval" to a constant?
